---
title: "2_Acro_Pocillo_UPGMA"
author: "Magena Marzonie"
date: "09/02/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

#Load libraries 
```{r}
library(ggsci)
library(tidyverse)
library(dplyr)
library(forcats)
library(reshape2)
library(stringr)
library(tidyr)
library(tibble)
library(sangerseqR)
library(DECIPHER)
library(Biostrings)
library(phangorn)
library(ape)
library(ggplot2)
library(ggtree)
library(patchwork)
library(bioseq)
library(kmer)
library(GUniFrac)
library(seqinr)
```

Custom functions (eventually move this to a separate file that we load in)
```{r}
dna_to_DNAbin <- function (dna){
  DNAbin <- as_DNAbin(dna)
  names(DNAbin) <- names(dna)
  return(DNAbin)
}

read_fasta_df <- function (file = "") {
  fasta <- readLines(file)
  ind <- grep(">", fasta)
  s <- data.frame(ind = ind, from = ind + 1, to = c((ind - 
    1)[-1], length(fasta)))
  seqs <- rep(NA, length(ind))
  for (i in 1:length(ind)) {
    seqs[i] <- paste(fasta[s$from[i]:s$to[i]], collapse = "")
  }
  tib <- tibble(label = gsub(">", "", fasta[ind]), sequence = seqs)
  return(tib)
}

dna_to_DNAbin <- function (dna){
  DNAbin <- as_DNAbin(dna)
  names(DNAbin) <- names(dna)
  return(DNAbin)
}

dna_to_DNAStringset <- function(x) 
{
    bioseq:::check_dna(x)
    DNAstr <- DNAStringSet(paste(x))
    names(DNAstr) <- names(x)
    return(DNAstr)
}
```

```{r}
#load sequence data from 1_ script 
load("Data/seqs_long.Rdata")
load("Data/full_meta.Rdata")
load("Data/all_data.Rdata")
load("Data/all_pal.Rdata")
load("Data/profile_data.Rdata")
```

```{r}
#load environmental data 
load("Data/SiteDisturbanceHistory_DHW.RData")

#add thermal history data to metadata 
full_meta2 = full_meta %>% 
  left_join(site.bleachings, by = "Site") %>% 
  mutate(DHW = case_when(DHW == "t" ~ "6.869999846", TRUE ~ DHW))

#relevel reef in order of latitude 
full_meta2 %>% mutate(Reef.x = factor(Reef.x, levels = c("Osprey", "Bougainville", "Moore", "Willis", "Holmes", "Chilcott", "Herald", "Lihou", "Flinders", "Marion", "Frederick", "Saumarez", "Wreck")))
```

#Three data frames for the Pocillopora species
```{r}
#keeping A.cf. humilis data - filtering out Poci and unknown samples 
poci_seqs <- seqs_long %>% filter(str_detect(Species, "P|k"))
pver_seqs <- seqs_long %>% filter(str_detect(Species, "Pverrucosa"))
pmea_seqs <- seqs_long %>% filter(str_detect(Species, "Pmeandrina"))
punk_seqs <- seqs_long %>% filter(str_detect(Species, "Unknown"))
```

# Fig 2B P. verrucosa Phylogenetic Tree

OBSERVATIONS: k-mer tree + unweighted unifrac AND the Alignment-tree + Gunifrac 0.5 do a good job of placing the southernmost samples together.

```{r}
#read in file 
fasta_pver <- read_fasta_df("20210612_marzonie/186_20211115_03_DBV_20211116T024440.seqs.fasta") %>%
  filter(label %in% pver_seqs$name) %>%   #only keeping DNA seqs that appear in seqs_long subset 
  filter(!str_detect(label, "A|G")) %>%
  deframe() %>%
  as_dna()

# Unifracs

## With a kmer-based tree

kdist_pver <- fasta_pver %>%
  dna_to_DNAbin() %>%
  kdistance(k = 7, residues = "DNA", method = "edgar") %>%
  as.matrix()

k_tree_pver <- kdist_pver %>% phangorn::upgma()

## With an alignment-based tree

fasta_dss <- fasta_pver %>%
  dna_to_DNAStringset()
aligned <- AlignSeqs(fasta_dss, verbose = TRUE)
writeXStringSet(aligned, filepath = "Data/verrucosa_aligned.fasta")
aligned <- read.alignment("Data/verrucosa_aligned.fasta", format = "fasta")
dist.al <- dist.alignment(aligned, gap = TRUE)
dist.al <- (as.matrix(dist.alignment(aligned, matrix = "identity" )))^2

align_tree_pver <- phangorn::upgma(dist.al)

# Get the community matrix

seqs_wide <- seqs_long %>%
  dplyr::select(plate_position, name, value) %>%
  filter(!str_detect(name, "A|G")) %>%
  filter(name %in% pver_seqs$name) %>%
  pivot_wider(names_from = name, values_from = value, values_fill = 0) %>%
  #mutate(across(everything(), replace_na, 0)) %>%
  filter(plate_position %in% pver_seqs$plate_position) %>%
  tibble::column_to_rownames(var = "plate_position")

pver.community.matrix = seqs_wide
save(pver.community.matrix, file = "Data/pver.community.matrix.RData")

# comput unifrac distances  
k_unidist_pver <- GUniFrac(seqs_wide, k_tree_pver)   #GUniFrac calculates all the distances 
k_unidist_pver <- k_unidist_pver$unifracs
align_unidist_pver <- GUniFrac(seqs_wide, align_tree_pver)   #GUniFrac calculates all the distances 
align_unidist_pver <- align_unidist_pver$unifracs

save(k_unidist_pver, file = "Data/unifracs_pver_kmer.RData")
save(align_unidist_pver, file = "Data/unifracs_pver_align.RData")

#Choose one for plotting
#du_pver <- k_unidist_pver[, , "d_UW"]    #unweighted unifrac
#du_pver <- align_unidist_pver[, , "d_UW"]    #unweighted unifrac
#du_pver <- k_unidist_pver[, , "d_1"]    # Weighted UniFrac
#du_pver <- align_unidist_pver[, , "d_1"]    # Weighted UniFrac
#du_pver <- k_unidist_pver[, , "d_0.5"]    # GUniFrac with alpha 0.5 
du_pver <- align_unidist_pver[, , "d_0.5"]    # GUniFrac with alpha 0.5 

# Cluster the samples
hclust_samps_pver <- upgma(du_pver)

# Get a sample order from ggtree
ggtree_samps_pver <- ggtree::ggtree(hclust_samps_pver)$data %>%
  filter(isTip == "TRUE") %>%
  arrange(y)

# Make the sample tree
tree_pver <- ggtree(hclust_samps_pver, size = 0.2) +
  scale_x_reverse() +
  theme(aspect.ratio = 1.8)

# add in metadata
tree_pver$data <- left_join(tree_pver$data, full_meta2, by = c("label" = "plate_position")) %>% 
  mutate(Reef.x = factor(Reef.x, levels = c("Osprey", "Bougainville", "Moore", "Willis", "Holmes", "Chilcott", "Herald", "Lihou", "Flinders", "Marion", "Frederick", "Saumarez", "Wreck")))

# color the trip tips
tree_tip_pver <- tree_pver + geom_tippoint(aes(color=-`GPS south`), size=2, alpha=1) +
    geom_tiplab(align=TRUE, size=1, linesize=.3, hjust = -3) +
  scale_color_viridis_c(option = "magma")

# Start plotting the composition data
plot_df_pver <- all_data %>%
  filter(str_detect(Species, "Pverrucosa")) %>%
  mutate(plate_position = fct_relevel(plate_position, ggtree_samps_pver$label))

theme_set(theme_bw())

bar_uni_pver <- 
ggplot(plot_df_pver, aes(value_rel, plate_position)) +
geom_bar(stat = "identity", aes(fill = name, colour = name)) +
theme(aspect.ratio = 0.5, legend.position = "none", axis.text.y=element_blank(), axis.ticks.y = element_blank(),
      axis.text.x=element_blank(), axis.ticks.x = element_blank(),
      axis.title.x = element_blank(), axis.title.y = element_blank(),
      panel.border = element_blank(), panel.grid.major = element_blank(),
panel.grid.minor = element_blank(), axis.ticks = element_blank()) +
scale_fill_manual(values = all_pal, breaks = levels(profile_data$name)) +
scale_colour_manual(values = all_pal, breaks = levels(profile_data$name)) +
geom_vline(xintercept = 1, size = 1) +
guides(fill=guide_legend(ncol=3))

#p_bar_uni is the sequences by colour. P_tree_tip is the tree coloured by reef. 

bar_uni_pver + tree_tip_pver
```

# Fig 2C P. meandrina Phylogenetic Tree

OBSERVATIONS: meandrina has a north/south split. This broadley holds true for all combinations of tree-type and unifrac-dists, but with minor differences between each method.

```{r}
#read in file 
fasta_pmea <- read_fasta_df("20210612_marzonie/186_20211115_03_DBV_20211116T024440.seqs.fasta") %>%
  filter(label %in% pmea_seqs$name) %>%   #only keeping DNA seqs that appear in seqs_long subset 
  filter(!str_detect(label, "A|G")) %>%
  deframe() %>%
  as_dna()

# Unifracs

## With a kmer-based tree

kdist_pmea <- fasta_pmea %>%
  dna_to_DNAbin() %>%
  kdistance(k = 7, residues = "DNA", method = "edgar") %>%
  as.matrix()

k_tree_pmea <- kdist_pmea %>% phangorn::upgma()

## With an alignment-based tree

fasta_dss <- fasta_pmea %>%
  dna_to_DNAStringset()
aligned <- AlignSeqs(fasta_dss, verbose = TRUE)
writeXStringSet(aligned, filepath = "Data/meandrina_aligned.fasta")
aligned <- read.alignment("Data/meandrina_aligned.fasta", format = "fasta")
dist.al <- dist.alignment(aligned, gap = TRUE)
dist.al <- (as.matrix(dist.alignment(aligned, matrix = "identity" )))^2

align_tree_mea <- phangorn::upgma(dist.al)

# Get the community matrix

seqs_wide <- seqs_long %>%
  dplyr::select(plate_position, name, value) %>%
  filter(!str_detect(name, "A|G")) %>%
  filter(name %in% pmea_seqs$name) %>%
  pivot_wider(names_from = name, values_from = value, values_fill = 0) %>%
  #mutate(across(everything(), replace_na, 0)) %>%
  filter(plate_position %in% pmea_seqs$plate_position) %>%
  tibble::column_to_rownames(var = "plate_position")

pmea.community.matrix = seqs_wide
save(pmea.community.matrix, file = "Data/pmea.community.matrix.RData")

# comput unifrac distances  
k_unidist_pmea <- GUniFrac(seqs_wide, k_tree_pmea)   #GUniFrac calculates all the distances 
k_unidist_pmea <- k_unidist_pmea$unifracs
align_unidist_pmea <- GUniFrac(seqs_wide, align_tree_mea)   #GUniFrac calculates all the distances 
align_unidist_pmea <- align_unidist_pmea$unifracs

# save unifracs for later use
save(k_unidist_pmea, file = "Data/unifracs_pmea_kmer.RData")
save(align_unidist_pmea, file = "Data/unifracs_pmea_align.RData")

#Choose one for plotting
du_pmea <- k_unidist_pmea[, , "d_UW"]    #unweighted unifrac
#du_pmea <- align_unidist_pmea[, , "d_UW"]    #unweighted unifrac
#du_pmea <- k_unidist_pmea[, , "d_1"]    # Weighted UniFrac
#du_pmea <- align_unidist_pmea[, , "d_1"]    # Weighted UniFrac
#du_pmea <- k_unidist_pmea[, , "d_0.5"]    # GUniFrac with alpha 0.5 
#du_pmea <- align_unidist_pmea[, , "d_0.5"]    # GUniFrac with alpha 0.5 

# Cluster the samples
hclust_samps_pmea <- upgma(du_pmea)

# Get a sample order from ggtree
ggtree_samps_pmea <- ggtree::ggtree(hclust_samps_pmea)$data %>%
  filter(isTip == "TRUE") %>%
  arrange(y)

# Make the sample tree
tree_pmea <- ggtree(hclust_samps_pmea, size = 0.2) +
  scale_x_reverse() +
  theme(aspect.ratio = 1.8)

# add in metadata
tree_pmea$data <- left_join(tree_pmea$data, full_meta2, by = c("label" = "plate_position")) %>% 
  mutate(Reef.x = factor(Reef.x, levels = c("Osprey", "Bougainville", "Moore", "Willis", "Holmes", "Chilcott", "Herald", "Lihou", "Flinders", "Marion", "Frederick", "Saumarez", "Wreck")))

# color the trip tips
tree_tip_pmea <- tree_pmea + geom_tippoint(aes(color=-`GPS south`), size=2, alpha=1) +
    geom_tiplab(align=TRUE, size=1, linesize=.3, hjust = -3) +
  scale_color_viridis_c(option = "magma")

# Start plotting the composition data
plot_df_pmea <- all_data %>%
  filter(str_detect(Species, "Pmeandrina")) %>%
  mutate(plate_position = fct_relevel(plate_position, ggtree_samps_pmea$label))

theme_set(theme_bw())

bar_uni_pmea <- 
ggplot(plot_df_pmea, aes(value_rel, plate_position)) +
geom_bar(stat = "identity", aes(fill = name, colour = name)) +
theme(aspect.ratio = 0.5, legend.position = "none", axis.text.y=element_blank(), axis.ticks.y = element_blank(),
      axis.text.x=element_blank(), axis.ticks.x = element_blank(),
      axis.title.x = element_blank(), axis.title.y = element_blank(),
      panel.border = element_blank(), panel.grid.major = element_blank(),
panel.grid.minor = element_blank(), axis.ticks = element_blank()) +
scale_fill_manual(values = all_pal, breaks = levels(profile_data$name)) +
scale_colour_manual(values = all_pal, breaks = levels(profile_data$name)) +
geom_vline(xintercept = 1, size = 1) +
guides(fill=guide_legend(ncol=3))

#p_bar_uni is the sequences by colour. P_tree_tip is the tree coloured by reef. 

bar_uni_pmea + tree_tip_pmea
```

# Fig 2C Poci unknown Phylogenetic Tree

OBSERVATIONS: as for meandrina, the unknown pocillo has a north/south split. This is even more consistent here than for meandrina.


```{r}
#read in file 
fasta_punk <- read_fasta_df("20210612_marzonie/186_20211115_03_DBV_20211116T024440.seqs.fasta") %>%
  filter(label %in% punk_seqs$name) %>%   #only keeping DNA seqs that appear in seqs_long subset 
  filter(!str_detect(label, "A|G")) %>%
  deframe() %>%
  as_dna()

# Unifracs

## With a kmer-based tree

kdist_punk <- fasta_punk %>%
  dna_to_DNAbin() %>%
  kdistance(k = 7, residues = "DNA", method = "edgar") %>%
  as.matrix()

k_tree_punk <- kdist_punk %>% phangorn::upgma()

## With an alignment-based tree

fasta_dss <- fasta_punk %>%
  dna_to_DNAStringset()
aligned <- AlignSeqs(fasta_dss, verbose = TRUE)
writeXStringSet(aligned, filepath = "Data/punkown_aligned.fasta")
aligned <- read.alignment("Data/punkown_aligned.fasta", format = "fasta")
dist.al <- dist.alignment(aligned, gap = TRUE)
dist.al <- (as.matrix(dist.alignment(aligned, matrix = "identity" )))^2

align_tree_unk <- phangorn::upgma(dist.al)

# Get the community matrix

seqs_wide <- seqs_long %>%
  dplyr::select(plate_position, name, value) %>%
  filter(!str_detect(name, "A|G")) %>%
  filter(name %in% punk_seqs$name) %>%
  pivot_wider(names_from = name, values_from = value, values_fill = 0) %>%
  #mutate(across(everything(), replace_na, 0)) %>%
  filter(plate_position %in% punk_seqs$plate_position) %>%
  tibble::column_to_rownames(var = "plate_position")

punk.community.matrix = seqs_wide
save(punk.community.matrix, file = "Data/punk.community.matrix.RData")

# comput unifrac distances  
k_unidist_punk <- GUniFrac(seqs_wide, k_tree_punk)   #GUniFrac calculates all the distances 
k_unidist_punk <- k_unidist_punk$unifracs
align_unidist_punk <- GUniFrac(seqs_wide, align_tree_unk)   #GUniFrac calculates all the distances 
align_unidist_punk <- align_unidist_punk$unifracs

save(k_unidist_punk, file = "Data/unifracs_punk_kmer.RData")
save(align_unidist_punk, file = "Data/unifracs_punk_align.RData")

#Choose one for plotting
#du_punk <- k_unidist_punk[, , "d_UW"]    #unweighted unifrac
#du_punk <- align_unidist_punk[, , "d_UW"]    #unweighted unifrac
#du_punk <- k_unidist_punk[, , "d_1"]    # Weighted UniFrac
#du_punk <- align_unidist_punk[, , "d_1"]    # Weighted UniFrac
#du_punk <- k_unidist_punk[, , "d_0.5"]    # GUniFrac with alpha 0.5 
du_punk <- align_unidist_punk[, , "d_0.5"]    # GUniFrac with alpha 0.5 

# Cluster the samples
hclust_samps_punk <- upgma(du_punk)

# Get a sample order from ggtree
ggtree_samps_punk <- ggtree::ggtree(hclust_samps_punk)$data %>%
  filter(isTip == "TRUE") %>%
  arrange(y)

# Make the sample tree
tree_punk <- ggtree(hclust_samps_punk, size = 0.2) +
  scale_x_reverse() +
  theme(aspect.ratio = 1.8)

# add in metadata
tree_punk$data <- left_join(tree_punk$data, full_meta2, by = c("label" = "plate_position")) %>% 
  mutate(Reef.x = factor(Reef.x, levels = c("Osprey", "Bougainville", "Moore", "Willis", "Holmes", "Chilcott", "Herald", "Lihou", "Flinders", "Marion", "Frederick", "Saumarez", "Wreck")))

# color the trip tips
tree_tip_punk <- tree_punk + geom_tippoint(aes(color=-`GPS south`), size=2, alpha=1) +
    geom_tiplab(align=TRUE, size=1, linesize=.3, hjust = -3) +
  scale_color_viridis_c(option = "magma")

# Start plotting the composition data
plot_df_punk <- all_data %>%
  filter(str_detect(Species, "Unknown")) %>%
  mutate(plate_position = fct_relevel(plate_position, ggtree_samps_punk$label))

theme_set(theme_bw())

bar_uni_punk <- 
ggplot(plot_df_punk, aes(value_rel, plate_position)) +
geom_bar(stat = "identity", aes(fill = name, colour = name)) +
theme(aspect.ratio = 0.5, legend.position = "none", axis.text.y=element_blank(), axis.ticks.y = element_blank(),
      axis.text.x=element_blank(), axis.ticks.x = element_blank(),
      axis.title.x = element_blank(), axis.title.y = element_blank(),
      panel.border = element_blank(), panel.grid.major = element_blank(),
panel.grid.minor = element_blank(), axis.ticks = element_blank()) +
scale_fill_manual(values = all_pal, breaks = levels(profile_data$name)) +
scale_colour_manual(values = all_pal, breaks = levels(profile_data$name)) +
geom_vline(xintercept = 1, size = 1) +
guides(fill=guide_legend(ncol=3))

#p_bar_uni is the sequences by colour. P_tree_tip is the tree coloured by reef. 

bar_uni_punk + tree_tip_punk
```

# Acropora humilis tree 

OBSERVATIONS: Regardless of the tree-type and unifrac combination, there is one major branch point in the tree. While this does not seem to be consistent with lat/lon, the unweighted approach seems very good at placing samples from the same lat/lons together. 
```{r}
#filtering seqs for acropora only 
acro_seqs <- seqs_long %>% filter(str_detect(Species, "A"))

#read in file 
fasta_acro <- read_fasta_df("20210612_marzonie/186_20211115_03_DBV_20211116T024440.seqs.fasta") %>%
  filter(label %in% acro_seqs$name) %>%   #only keeping DNA seqs that appear in seqs_long subset 
  filter(!str_detect(label, "A|G")) %>%
  deframe() %>%
  as_dna()

# Unifracs

## With a kmer-based tree

kdist_acro <- fasta_acro %>%
  dna_to_DNAbin() %>%
  kdistance(k = 7, residues = "DNA", method = "edgar") %>%
  as.matrix()

k_tree_acro <- kdist_acro %>% phangorn::upgma()

## With an alignment-based tree

fasta_dss <- fasta_acro %>%
  dna_to_DNAStringset()
aligned <- AlignSeqs(fasta_dss, verbose = TRUE)
writeXStringSet(aligned, filepath = "Data/acro_aligned.fasta")
aligned <- read.alignment("Data/acro_aligned.fasta", format = "fasta")
dist.al <- dist.alignment(aligned, gap = TRUE)
dist.al <- (as.matrix(dist.alignment(aligned, matrix = "identity" )))^2

align_tree_acro <- phangorn::upgma(dist.al)

# Get the community matrix

seqs_wide <- seqs_long %>%
  dplyr::select(plate_position, name, value) %>%
  filter(!str_detect(name, "A|G")) %>%
  filter(name %in% acro_seqs$name) %>%
  pivot_wider(names_from = name, values_from = value, values_fill = 0) %>%
  #mutate(across(everything(), replace_na, 0)) %>%
  filter(plate_position %in% acro_seqs$plate_position) %>%
  tibble::column_to_rownames(var = "plate_position")

acro.community.matrix = seqs_wide
save(acro.community.matrix, file = "Data/acro.community.matrix.RData")

# comput unifrac distances  
k_unidist_acro <- GUniFrac(seqs_wide, k_tree_acro)   #GUniFrac calculates all the distances 
k_unidist_acro <- k_unidist_acro$unifracs
align_unidist_acro <- GUniFrac(seqs_wide, align_tree_acro)   #GUniFrac calculates all the distances 
align_unidist_acro <- align_unidist_acro$unifracs

save(k_unidist_acro, file = "Data/unifracs_acro_kmer.RData")
save(align_unidist_acro, file = "Data/unifracs_acro_align.RData")

#Choose one for plotting
#du_acro <- k_unidist_acro[, , "d_UW"]    #unweighted unifrac
#du_acro <- align_unidist_acro[, , "d_UW"]    #unweighted unifrac
#du_acro <- k_unidist_acro[, , "d_1"]    # Weighted UniFrac
#du_acro <- align_unidist_acro[, , "d_1"]    # Weighted UniFrac
#du_acro <- k_unidist_acro[, , "d_0.5"]    # GUniFrac with alpha 0.5 
du_acro <- align_unidist_acro[, , "d_0.5"]    # GUniFrac with alpha 0.5 

# Cluster the samples
hclust_samps_acro <- upgma(du_acro)

# Get a sample order from ggtree
ggtree_samps_acro <- ggtree::ggtree(hclust_samps_acro)$data %>%
  filter(isTip == "TRUE") %>%
  arrange(y)

# Make the sample tree
tree_acro <- ggtree(hclust_samps_acro, size = 0.2) +
  scale_x_reverse() +
  theme(aspect.ratio = 1.8)

# add in metadata
tree_acro$data <- left_join(tree_acro$data, full_meta2, by = c("label" = "plate_position")) %>% 
  mutate(Reef.x = factor(Reef.x, levels = c("Osprey", "Bougainville", "Moore", "Willis", "Holmes", "Chilcott", "Herald", "Lihou", "Flinders", "Marion", "Frederick", "Saumarez", "Wreck")))

# color the trip tips
tree_tip_acro <- tree_acro + geom_tippoint(aes(color=-`GPS south`), size=2, alpha=1) +
    geom_tiplab(align=TRUE, size=1, linesize=.3, hjust = -3) +
  scale_color_viridis_c(option = "magma")

# Start plotting the composition data
plot_df_acro <- all_data %>%
  filter(str_detect(Species, "Ahumilis")) %>%
  mutate(plate_position = fct_relevel(plate_position, ggtree_samps_acro$label))

theme_set(theme_bw())

bar_uni_acro <- 
ggplot(plot_df_acro, aes(value_rel, plate_position)) +
geom_bar(stat = "identity", aes(fill = name, colour = name)) +
theme(aspect.ratio = 0.7, legend.position = "none", axis.text.y=element_blank(), axis.ticks.y = element_blank(),
      axis.text.x=element_blank(), axis.ticks.x = element_blank(),
      axis.title.x = element_blank(), axis.title.y = element_blank(),
      panel.border = element_blank(), panel.grid.major = element_blank(),
panel.grid.minor = element_blank(), axis.ticks = element_blank()) +
scale_fill_manual(values = all_pal, breaks = levels(profile_data$name)) +
scale_colour_manual(values = all_pal, breaks = levels(profile_data$name)) +
geom_vline(xintercept = 1, size = 1) +
guides(fill=guide_legend(ncol=3))

#p_bar_uni is the sequences by colour. P_tree_tip is the tree coloured by reef. 

bar_uni_acro + tree_tip_acro
```
#Saving objects for script 3_PCA
```{r}
save(full_meta, file = "Data/full_meta.Rdata")
save(full_meta2, file = "Data/full_meta2.Rdata")
```

